src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt

```diff
@@ -0,0 +1,71 @@
+package uk.gov.dluhc.printapi.aspect
+
+import org.aspectj.lang.JoinPoint
+import org.aspectj.lang.annotation.AfterReturning
+import org.aspectj.lang.annotation.Aspect
+import org.springframework.stereotype.Component
+import uk.gov.dluhc.printapi.database.entity.AnonymousElectorDocument
+import uk.gov.dluhc.printapi.database.entity.Certificate
+import uk.gov.dluhc.printapi.database.entity.TemporaryCertificate
+import uk.gov.dluhc.printapi.service.StatisticsUpdateService
+
+@Aspect
+@Component
+class StatisticsUpdateAspect(
+    private val statisticsUpdateService: StatisticsUpdateService
+) {
+    @AfterReturning(
+        pointcut = "execution(* org.springframework.data.repository.CrudRepository.save(..)) && args(certificate)",
+        returning = "saved"
+    )
+    fun afterSaveCertificate(joinPoint: JoinPoint, certificate: Certificate, saved: Certificate) {
+        saved.sourceReference?.also {
+            statisticsUpdateService.triggerVoterCardStatisticsUpdate(it)
+        }
+    }
+
+    @AfterReturning(
+        pointcut = "execution(* org.springframework.data.repository.CrudRepository.save(..)) && args(temporaryCertificate)",
+        returning = "saved"
+    )
+    fun afterSaveTemporaryCertificate(joinPoint: JoinPoint, temporaryCertificate: TemporaryCertificate, saved: TemporaryCertificate) {
+        saved.sourceReference?.also {
+            statisticsUpdateService.triggerVoterCardStatisticsUpdate(it)
+        }
+    }
+
+    @AfterReturning(
+        pointcut = "execution(* org.springframework.data.repository.CrudRepository.save(..)) && args(anonymousElectorDocument)",
+        returning = "saved"
+    )
+    fun afterSaveTemporaryCertificate(joinPoint: JoinPoint, anonymousElectorDocument: AnonymousElectorDocument, saved: AnonymousElectorDocument) {
+        statisticsUpdateService.triggerVoterCardStatisticsUpdate(saved.sourceReference)
+    }
+
+    @AfterReturning(
+        pointcut = "execution(* org.springframework.data.repository.CrudRepository.saveAll(..))",
+        returning = "saved"
+    )
+    fun afterSaveAll(joinPoint: JoinPoint, saved: Iterable<Any>) {
+        // Type is erased and AOP is passed an iterable containing any type
```

Comment:
I'm not sure I understand this - what's AOP? What type is erased?

Reply:
Hmm this comment came over from VCA where I ~copied~ got inspiration for this code from. AOP is "Aspect Oriented Programming", the comment is saying that while the `afterSave` methods have explicit types in the signatures this one has an `Any` type, so we need to check types explicitly. Honestly I don't think the comment adds anything - happy to remove?

Reply:
Is it worth maybe mentioning how we know all objects are of the same type?